core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1178,6 +1178,37 @@ abstract class RDD[T: ClassTag](
     }
   }
 
+  /**
+   * Computing desired tree aggregate depth necessary to avoid exceeding
+   * driver.MaxResultSize during aggregation.
+   * Based on the formulae: (numPartitions)^(1/depth) * objectSize <= DriverMaxResultSize
+   * @param aggregatedObjectSizeInMb the size, in megabytes, of the object being tree aggregated
+   * @param numPartitions the number of partitions for which to aggregate partial results
+   * @param maxDriverResultSizeInMb the value of the parameter spark.driver.maxResultSize in MB
+   */
+  def getTreeAggregateIdealDepth(
+      aggregatedObjectSizeInMb: Int,
+      numPartitions: Int = getNumPartitions,
+      maxDriverResultSizeInMb: Int = Utils.memoryStringToMb(
+        context.getConf.get("spark.driver.maxResultSize"))): Int = {
+    require(
+      aggregatedObjectSizeInMb < maxDriverResultSizeInMb,
+      s"object size to aggregate ($aggregatedObjectSizeInMb MB) exceeds max driver result size"
+        + s"($maxDriverResultSizeInMb MB)"
+    )
+    val numerator = math.log10(numPartitions) + math.log10(aggregatedObjectSizeInMb)
+    val denominator = math.log10(maxDriverResultSizeInMb)
+    val desiredTreeDepth = math.ceil(numerator / denominator).toInt
```

Comment:
I might be doing the math wrong here, but we want partitions^(1/depth) * vectorSize <= maxResultSize, right? I get that this needs depth >= log(partitions) / (log(maxResultSize) - log(vectorSize)) ?

I suppose we want some safety checks for 0 partitions or 0 aggregate size in MB.

Reply:
You're right, the formulae you post is actually what I tested at scale. Dunno why I didn't just copy paste it, I just brain farted on the maths when writing the PR ^^